### PR TITLE
Module 0 complete

### PR DESCRIPTION
**Pull request by srodriguez-lefebre**
Check the last exercises

### CODE REVIEW

File: 0_FirstModule/0_torch test.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+print("Tensor not on GPU")
+print(tensor, tensor.device)
+
+print("Move tensor to GPU (if available)")
+tensor_on_gpu = tensor.to(device)
+print(tensor_on_gpu)
+
+print("Notice the second tensor has device='cuda:0', this means it's stored on the 0th GPU available (GPUs are 0 indexed, if two GPUs were available, they'd be 'cuda:0' and 'cuda:1' respectively, up to 'cuda:n').")  
+
+print("If tensor is on GPU, can't transform it to NumPy (this will error)")
```
Comment (JoaquinCampo) [inline 0_FirstModule/0_torch test.py:27]: Just out of curiosity, why wouldnt you be able to transform a tensor to a numpy array if it is stored on the gpu?

File: 0_FirstModule/1_Manipulating tensors (tensor operations).py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+print(tensor01 * tensor01)
+
+print("Matrix multiplication")
+print(torch.matmul(tensor01, tensor01))
+
+print("Can also use the '@' symbol for matrix multiplication, though not recommended")
+print(tensor01 @ tensor01)
+
+print("Matrix multiplication by hand")
+print("(avoid doing operations with for loops at all cost, they are computationally expensive)")
```
Comment (JoaquinCampo) [inline 0_FirstModule/1_Manipulating tensors (tensor operations).py:120]: Why are loops expensive? Why is it advised to do matrix multiplications instead of for loops?

File: 0_FirstModule/3_Indexing & NumPy & Reproducibility.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+print(f"Tensor A:\n{random_tensor_A}\n")
+print(f"Tensor B:\n{random_tensor_B}\n")
+print(f"Does Tensor A equal Tensor B? (anywhere)")
+print(random_tensor_A == random_tensor_B)
+
+import torch
+import random
+
+print("Set the random seed")
+RANDOM_SEED=42 # try changing this to different values and see what happens to the numbers below
```
Comment (JoaquinCampo) [inline 0_FirstModule/3_Indexing & NumPy & Reproducibility.py:54]: why is the seed importante? Why dont just use random without a seed?

**Review by JoaquinCampo — COMMENTED** [review]
Good job! Keep up the enthusiasm, this is just the beginning!

**Review by JoaquinCampo — APPROVED** [review]

File: 0_FirstModule/0_torch test.py (hunk 1 above)
Comment (srodriguez-lefebre) [reply 0_FirstModule/0_torch test.py:27]: This is a good question, because I didn't pay attention to this comment and I realized that I didn't understand the reason why. So I looked for it and I found it:
Because NumPy operates only on CPU memory. When a tensor is on the GPU, it resides in the GPU's VRAM (Video Random Access Memory), which is separate from the CPU's memory where NumPy operates.

File: 0_FirstModule/3_Indexing & NumPy & Reproducibility.py (hunk 3 above)
Comment (srodriguez-lefebre) [reply 0_FirstModule/3_Indexing & NumPy & Reproducibility.py:54]: Check this because the following sentence is what I think and don't look for this.
 I think the reason of setting a seed is for "Reproducibility" for example when you find a good result and you want share with someone, you need the "random values" but you want the same result so you use the same seed that you get the good result

File: 0_FirstModule/1_Manipulating tensors (tensor operations).py (hunk 2 above)
Comment (srodriguez-lefebre) [reply 0_FirstModule/1_Manipulating tensors (tensor operations).py:120]: For two principal reasons, i think:
1- Performance: In PyTorch (and other numerical computing libraries), operations are optimized to work on entire tensors at once.
2- Speed: Tensor operations in PyTorch are implemented in C++ and are highly optimized.
Additionally, when using a GPU:
3- GPU Utilization: When using a GPU, tensor operations can be parallelized across thousands of cores. For loops in Python cannot efficiently utilize GPU parallelism.
